### DISCUSSION TITLE
Why is the property EqualityContract on a C# record virtual?

### DISCUSSION THREAD

**Discussion by bitbonk** [dotnet/csharplang#6951] [Q&A]
The record type of C# 9 has the following property 

```csharp
Type EqualityContract { get; }
```

If the record is not sealed, [that property will be virtual](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/records#equality-members) so it can be overridden.
Unfortunately the docs do not state what I would use this property for.  Why would I ever want to override this property? Can you give me a somewhat realistic example?

I tried to override the `EqualityContract` property to allow equality of derived types but it doesn't work. This raised the question, if overriding it can't be used for that, what _can_ it be used for?

```csharp
[Fact]
public void Test()
{
    Foo foo = new Foo("name");
    Foo fooDerived = new FooDerived("name");
    Assert.True(foo.Equals(fooDerived));
    Assert.True(fooDerived.Equals(foo)); // fails
}

record Foo(string Name)
{
    public virtual bool Equals(Foo? other)
    {
        return other != null && this.Name.Equals(other.Name);
    }
    public override int GetHashCode()
    {
        return this.Name.GetHashCode();
    }
    protected virtual Type EqualityContract { get; } = typeof(Foo);
}

record FooDerived(string Name) : Foo(Name)
{
    public virtual bool Equals(FooDerived? other)
    {
        return other != null && this.Name.Equals(other.Name);
    }
    public override int GetHashCode()
    {
        return this.Name.GetHashCode();
    }
    protected override Type EqualityContract { get; } = typeof(Foo);
}
```

**Comment by yaakov-h** [comment]
Typically you wouldn't override it, but the compiler would. In the compiler-generated `Equals` function, it signals to the base class that the subclass is not of the same type, so that `base.Equals(derived)` knows to return `false` even if all of the members in the base class are identical in both object instances.

**Reply by svick** [reply]
In particular consider these simple records:

```c#
record Foo(string Name);

record FooDerived(string Name) : Foo(Name);
```

The [generated code](https://sharplab.io/#v2:D4Jwpgxg9iAmAEAxKUAUABAjABngOQEMBbMASgG4BYAKBvGjiRQBEwQBLANzFgx32Jl4ALiZpCJCkA==) for them is (edited for brevity):

```c#
class Foo : IEquatable<Foo>
{
    private readonly string <Name>k__BackingField;

    protected virtual Type EqualityContract => typeof(Foo);

    public string Name
    {
        get => <Name>k__BackingField;
        init => <Name>k__BackingField = value;
    }

    public Foo(string Name)
    {
        <Name>k__BackingField = Name;
    }

    public override bool Equals(object obj) => Equals(obj as Foo);

    public virtual bool Equals(Foo other)
    {
        if ((object)this != other)
        {
            if ((object)other != null && EqualityContract == other.EqualityContract)
            {
                return EqualityComparer<string>.Default.Equals(<Name>k__BackingField, other.<Name>k__BackingField);
            }
            return false;
        }
        return true;
    }
}

internal class FooDerived : Foo, IEquatable<FooDerived>
{
    protected override Type EqualityContract => typeof(FooDerived);

    public FooDerived(string Name) : base(Name) { }

    public override bool Equals(object obj) => Equals(obj as FooDerived);

    public sealed override bool Equals(Foo other) => Equals((object)other);

    public virtual bool Equals(FooDerived other)
    {
        if ((object)this != other)
        {
            return base.Equals(other);
        }
        return true;
    }
}
```

Basically, `EqualityContract` exists to ensure that `new Foo("name").Equals(new FooDerived("name"))` is `false`.

**Reply by FaustVX** [reply]
What is the need for `EqualityContract` instead of simple `.GetType()`:
`if ((object)other != null && GetType() == other.GetType())`

**Reply by CyrusNajmabadi** [reply]
To allow you to override the behavior if you want.  You can choose for multiple records in an inheritance tree to be considered equal even if they aren't the same type.

**Reply by yaakov-h** [reply]
https://github.com/dotnet/csharplang/issues/3137#issuecomment-581254766

**Reply by bitbonk** [reply]
I understand that the property `Type EqualityContract { get; }` is needed by the compiler. But that doesn't explain why it has been made _virtual_. It seems that no one except the compiler itself can _ever_ do something useful with overriding the virtual property.  This seems like a leaky abstraction in the feature design. 

@yaakov-h 
> Typically you wouldn't override it

Is there a atypical, yet valid reason to override it in user code?

**Reply by bernd5** [reply]
If it wouldn't be virtual the compiler could not override it....

**Reply by bitbonk** [reply]
I suspect there would have been other means to achieve the equality contract functionality with the usual code generation magic of the C# compiler without introducing a virtual property, overridable by the user.

**Reply by CyrusNajmabadi** [reply]
>  But that doesn't explain why it has been made virtual

So it can be overridden.  

> Is there a atypical, yet valid reason to override it in user code?

See https://github.com/dotnet/csharplang/issues/3137#issuecomment-581254766

**Reply by CyrusNajmabadi** [reply]
> I suspect there would have been other means to achieve the equality contract 

Sure.  There were probably other routes we could have taken.  But why would we?  This route was simple and works very well.

**Reply by bitbonk** [reply]
@CyrusNajmabadi 

> See #3137 (comment)

I don't think it is possible to override that property in a meaningful way _in user code_. At least I can't come up with an example. Do you have an example?

**Reply by alexrp** [reply]
```csharp
record Base();

record Derived : Base
{
    protected override Type EqualityContract => typeof(Base);
}
```

?

**Reply by CyrusNajmabadi** [reply]
@bitbonk The comment l linked to gives an example.

**Reply by bitbonk** [reply]
@alexrp 
> ```cs
> record Base();
> 
> record Derived : Base
> {
>     protected override Type EqualityContract => typeof(Base);
> }
> ```
> 
> ?

Thank you! 🙏
Upon further investigation of this simple example, I finally understood what I was misunderstanding here.

That example still makes my test fail:
```cs
[Fact]
public void Test2()
{
    Base @base = new Base();
    Base derived = new Derived();
    Assert.True(@base.Equals(derived));
    Assert.True(derived.Equals(@base)); // fails here
}
```
As a result, I was wrongly assuming that overriding the `EqualityContract` property has no effect. But it actually does: If I remove the override
```cs
record Base;
record Derived : Base;
``` 
the test already fails in the first assertion.
```cs
[Fact]
public void Test2()
{
    Base @base = new Base();
    Base derived = new Derived();
    Assert.True(@base.Equals(derived)); // fails here
    Assert.True(derived.Equals(@base)); 
}
```
So overriding the property does makes sense if you want base and derived considered to be equal. It is just important to keep in mind that equality is not symmetrical. For records, it is generally not possible to make `derived.Equals(@base)` return `true`.  This is a bit suprising.
